maze_transformer/training/dataset.py

```diff
@@ -1,23 +1,45 @@
-from dataclasses import dataclass, field
 from functools import cached_property
 
 import numpy as np
 import torch
-from muutils.tensor_utils import ATensor
+from muutils.json_serialize import (
+    SerializableDataclass,
+    serializable_dataclass,
+    serializable_field,
+)
+from muutils.tensor_utils import DTYPE_MAP, ATensor
 from torch.utils.data import Dataset
 
 from maze_transformer.utils.utils import get_device
 
 
-@dataclass(kw_only=True)
-class GPTDatasetConfig:
+def _dtype_serialization_fn(x: torch.dtype | np.dtype) -> str:
+    x_str: str = str(x)
+    assert x_str in DTYPE_MAP, f"unknown dtype {x}"
+    assert DTYPE_MAP[x_str] == x
+    return x_str
+
+
+@serializable_dataclass(
+    kw_only=True,
+    properties_to_serialize=["token_arr", "padding_token_index", "tokenizer_map"],
+)
+class GPTDatasetConfig(SerializableDataclass):
     """base config class"""
 
     name: str
-    device: torch.device = field(default_factory=lambda: torch.device(get_device()))
-    dtype: torch.dtype | np.dtype = field(default_factory=lambda: torch.int16)
-    seq_len_min: int = 1
-    seq_len_max: int = 512
+    device: torch.device = serializable_field(
+        default_factory=lambda: torch.device(get_device()),
+        serialization_fn=lambda x: str(x),
```

Comment:
I need to clarify this in the ZANJ docs, but in brief:

Given a class
```python
@serializable_dataclass
class MyClass(SerializableDataclass):
    item1: str
    item2: dict[str, int]
    item3: SomeClass
    item4: list[SomeClass]
```

where `SomeClass` is either a `SerializableDataclass` or implements a compatible interface, all items except `item4` will always work out of the box. Because `item4` is neither serializable to json by default, and nor is it an instance of a `SerializableDataclass` itself, you have to write the compatibility layer in `serialization_fn`. Hopefully this explains the comment

You also have to implement a `serialization_fn` and `loading_fn` if the type of your class member is something other than a base type or `SerializableDataclass`. For things like a `torch.dtype` or `torch.optimizer` there isn't really a way around this except pickling. The `SerializableDataclass` and `serializable_field` interface hopefully provides a relatively painless way of doing this.

Reply:
Gotcha, makes sense!